Zend/zend_hash.h

```diff
@@ -1149,6 +1149,33 @@ static zend_always_inline void *zend_hash_get_current_data_ptr_ex(HashTable *ht,
 		ZEND_HASH_FILL_FINISH(); \
 	} while (0)
 
+/* Check if an array is a list */
+static zend_always_inline zend_bool zend_array_is_list(zend_array *array)
+{
+	zend_long expected_idx = 0;
+	zend_long num_idx;
+	zend_string* str_idx;
+	/* Empty arrays are lists */
+	if (zend_hash_num_elements(array) == 0) {
+		return 1;
+	}
+
+	/* Packed arrays are lists */
+	if (HT_IS_PACKED(array) && HT_IS_WITHOUT_HOLES(array)) {
+		return 1;
+	}
+
+	/* Check if the list could theoretically be repacked */
+	ZEND_HASH_FOREACH_KEY(array, num_idx, str_idx) {
```

Comment:
Could we rebase it to master, and try to use `ZEND_HASH_REVERSE_FOREACH_KEY`, just like what you said on https://github.com/php/php-src/pull/4886#issuecomment-551125929, Thanks!